core/src/main/java/tech/tablesaw/api/Table.java

```diff
@@ -1135,6 +1136,118 @@ public boolean hasNext() {
+   * label | value1 | value2 |<br>
+   * -------------------------------<br>
+   * row1 | 1 | 2 |<br>
+   * row2 | 1.1 | 2.1 |<br>
+   * row3 | 1.2 | 2.2 |<br>
+   *
+   * <p>Is transposed into the following<br>
+   * label | row1 | row2 | row3 |<br>
+   * -------------------------------------<br>
+   * value1 | 1 | 1.1 | 1.2 |<br>
+   * value2 | 2 | 2.1 | 2.2 |<br>
+   *
+   * @param includeColumnHeadingsAsFirstColumn Toggle whether to include the column headings as
+   *     first column in result
+   * @param useFirstColumnForHeadings Use the first column as the column headings in the result.
+   *     Useful if the data set already has a first column which contains a set of labels
+   * @return The transposed table
+   */
+  public Table transpose(
+      boolean includeColumnHeadingsAsFirstColumn, boolean useFirstColumnForHeadings) {
+    if (this.columnCount() == 0) {
+      return this;
+    }
+
+    // Validate first
+    int columnOffset = useFirstColumnForHeadings ? 1 : 0;
+    ColumnType resultColumnType = validateTableHasSingleColumnType(columnOffset);
+
+    Table transposed = Table.create(this.name);
+    if (includeColumnHeadingsAsFirstColumn) {
+      String columnName = useFirstColumnForHeadings ? this.column(0).name() : "0";
+      StringColumn labelColumn = StringColumn.create(columnName);
+      for (int i = columnOffset; i < this.columnCount(); i++) {
+        Column<?> columnToTranspose = this.column(i);
+        labelColumn.append(columnToTranspose.name());
+      }
+      transposed.addColumns(labelColumn);
+    }
+
+    if (useFirstColumnForHeadings) {
+      transpose(transposed, resultColumnType, row -> String.valueOf(this.get(row, 0)), 1);
+    } else {
+      // Use defaults
+      transpose(transposed, resultColumnType);
+    }
+    return transposed;
+  }
+
+  private ColumnType validateTableHasSingleColumnType(int startingColumn) {
+    // If all columns are of the same type
+    ColumnType[] columnTypes = this.columnTypes();
+    long distinctColumnTypesCount =
+        Arrays.stream(columnTypes).skip(startingColumn).distinct().count();
+    if (distinctColumnTypesCount > 1) {
+      throw new IllegalArgumentException(
+          "This operation currently only supports tables where value columns are of the same type");
+    }
+    return columnTypes[startingColumn];
+  }
+
+  private Table transpose(Table transposed, ColumnType resultColumnType) {
+    // default column labelling
+    return transpose(
+        transposed, resultColumnType, row -> String.valueOf(transposed.columnCount()), 0);
+  }
+
+  private Table transpose(
+      Table transposed,
+      ColumnType resultColumnType,
+      IntFunction<String> columnNameExtractor,
+      int startingColumn) {
+
+    for (int row = 0; row < this.rowCount(); row++) {
+      String columnName = columnNameExtractor.apply(row);
+      Column column = resultColumnType.create(columnName);
```

Comment:
This will result in a compilation warning. You should either address or suppress

Reply:
I had to remove the <?> so that I could make use of append(). 
This issue could be fixed by fixing the generics on the ColumnType class. ColumnType.create() shouldn't really be returning <?>, instead it could be typed to return <T> as the actual column type. 
I started looking at that over the weekend, but backed out of it because it was going to be quite a big refactor.